Playwright Conversion from Selenium

cmsato09 opened the pull request:
First pass at changing selenium to playwright
Got it to work with a simple carbon command `carbon -s 'print("Hi World")'`

- commented out selenium functions for now
- headless=False, so it's easy to confirm it was opening chrome and opening url page
- "carbon_image.png" should save in the set destination folder
- `wait_for_event()` let's the browser finish downloading before closing. (may not need that `sleep` function)

Currently fails following pytests:

- test_create_image_for_one_liner
- test_create_image_for_larger_snippet
- test_storing_image_in_different_folder

bbelderbos commented on carbon/carbon.py:
```diff
@@ -37,27 +38,47 @@ def _create_carbon_url(code, **carbon_options: str) -> str:
-        driver.find_element(By.ID, "export-png").click()
-        # make sure it has time to download the image
+    with sync_playwright() as p:
+        browser = p.chromium.launch(headless=False)
```
With Option going away (no more Selenium) what about making this an environment variable, I would make this configurable.

bbelderbos commented on carbon/carbon.py:
```diff
@@ -3,10 +3,11 @@
-from selenium.webdriver.chrome.service import Service
-from selenium.webdriver.common.by import By
+from playwright.sync_api import sync_playwright
+# from selenium import webdriver
```
You can just delete commented code

bbelderbos commented on pyproject.toml:
```diff
@@ -11,6 +11,7 @@ classifiers = ["License :: OSI Approved :: MIT License"]
     "pyperclip==1.8.2",
+    "pytest-playwright>=0.3.3",
     "python-dotenv==0.21.0",
     "selenium==4.6.1",
```
goal is to remove this

bbelderbos commented on pyproject.toml:
```diff
@@ -11,6 +11,7 @@ classifiers = ["License :: OSI Approved :: MIT License"]
 dynamic = ["version", "description"]
 dependencies = [
     "pyperclip==1.8.2",
+    "pytest-playwright>=0.3.3",
```
is py still set to >=3.7? we might want to make this a more recent min requirement

it would also be nice to use uv from here on

bbelderbos replied:
And use the new `uv publish`, let's try it :)

cmsato09 commented:
- Updated pyproject.toml file with correct dependencies
  - moved mypy, pytest-playwright, ruff to dev-dependencies
  - ruff linting config settings --> includes ignore E501 
  - mypy config settings --> ignore_missing_import
- Removed code related to Selenium
- Moved carbon code image generation logic into `with` statement

cmsato09 commented:
Add ruff to pre-commit #15 

- copied ruff pre-commit hook as seen in documentation. Lint fixes enabled.
https://docs.astral.sh/ruff/integrations/#pre-commit

bbelderbos commented on carbon/carbon.py:
```diff
@@ -38,47 +34,21 @@ def _create_carbon_url(code, **carbon_options: str) -> str:
+            download_path = os.path.join(destination, "carbon_image.png")
+            download.save_as(download_path)
+            
+            sleep(SECONDS_SLEEP_BEFORE_DOWNLOAD)
```
do we still need this to be configurable? I think a hardcoded value would be fine for now?

bbelderbos commented on carbon/carbon.py:
```diff
@@ -38,47 +34,21 @@ def _create_carbon_url(code, **carbon_options: str) -> str:
     destination = kwargs.get("destination", os.getcwd())
-    prefs = {"download.default_directory": destination}
-    # options.add_experimental_option("prefs", prefs)
+    headless = kwargs.get("headless", True)
```
I agree, should be headless by default, bug what about controlling this with a DEBUG variable? so we can run it with headless=False if we want to see what it's doing in development / locally?

cmsato09 replied:
Not sure, but I think it's related to the `--interactive` flag when it was still on Selenium. 
Need to figure out what to do with `--interactive` in cli.py

bbelderbos replied:
What about the DEBUG env variable suggestion?

bbelderbos reviewed (approved):
Happy with the PR, just 2 follow up questions, thanks for this effort, really happy with the changes.

cmsato09 commented on requirements.txt:
```diff
@@ -1,3 +1,14 @@
-selenium==4.11
+# This file was autogenerated by uv via the following command:
+#    uv pip compile pyproject.toml -o requirements.txt
+greenlet==3.0.3
```
Just in case people haven't switched to poetry or uv and still using pip. 
Haven't checked to see if you can install dependencies from pyproject.toml with pip.
